core/src/main/java/com/linecorp/armeria/internal/client/AbstractRuleBuilderUtil.java

```diff
@@ -41,9 +42,42 @@ public final class AbstractRuleBuilderUtil {
     buildFilter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
                 @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter,
                 @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter,
+                @Nullable BiPredicate<ClientRequestContext, HttpHeaders> grpcTrailersFilter,
                 @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter,
                 boolean hasResponseFilter) {
-        return (ctx, cause) -> {
+
+        return new Filter(requestHeadersFilter, exceptionFilter, responseHeadersFilter,
+                          responseTrailersFilter, grpcTrailersFilter, hasResponseFilter);
+    }
+
+    private AbstractRuleBuilderUtil() {}
+
+    private static class Filter implements BiFunction<ClientRequestContext, Throwable, Boolean> {
+        private final BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter;
+        private final @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter;
+        private final @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter;
+        private final @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter;
+        private final @Nullable BiPredicate<ClientRequestContext, HttpHeaders> grpcTrailersFilter;
+        private final boolean hasResponseFilter;
+
+        Filter(BiPredicate<ClientRequestContext, RequestHeaders> requestHeadersFilter,
+               @Nullable BiPredicate<ClientRequestContext, Throwable> exceptionFilter,
+               @Nullable BiPredicate<ClientRequestContext, ResponseHeaders> responseHeadersFilter,
+               @Nullable BiPredicate<ClientRequestContext, HttpHeaders> responseTrailersFilter,
+               @Nullable BiPredicate<ClientRequestContext, HttpHeaders> grpcTrailersFilter,
+               boolean hasResponseFilter) {
+            this.requestHeadersFilter = requestHeadersFilter;
+            this.exceptionFilter = exceptionFilter;
+            this.responseHeadersFilter = responseHeadersFilter;
+            this.responseTrailersFilter = responseTrailersFilter;
+            this.grpcTrailersFilter = grpcTrailersFilter;
+            this.hasResponseFilter = hasResponseFilter;
+        }
+
+        // TODO(trustin): Figure out why apply() is called more than once for each client call
```

Comment:
Which test case makes calls this more than once?

Reply:
I investigated again and it was actually that the client was making two calls, i.e. there's no problem actually. Removing the comment.